[javascript] CodeQL query to detect  if cookies are sent without the flag secure being set

dellalibera opened the pull request:
Query to detect if the `secure` flag is set to cookies is available in java query but it is not available in javascript query.

This query detects if cookies are sent without the flag `secure` being set or with the flag `secure` being set to `false`.

Failing to set the `secure` flag on a cookie can cause it to be sent in cleartext.
This makes it easier for an attacker to intercept and read the cookie.

This query handles the cases when cookies are set with:
- Express middleware `cookie-session` (https://github.com/expressjs/cookie-session)
- Express middleware `express-session` (https://github.com/expressjs/session)
- Express `response.cookie` (https://expressjs.com/en/api.html#res.cookie)
- `Set-Cookie` of an HTTP response (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie)
- `js-cookie` (https://github.com/js-cookie/js-cookie)

I added several examples showing the different cases when the cookie is sent without the flag `secure` being set or with the flag `secure` being set to `false`.

Any feedback is welcome.
I hope this will help.

intrigus-lgtm commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qhelp:
```diff
@@ -8,7 +8,7 @@ This makes it easier for an attacker to intercept.</p>
 <recommendation>
 
-<p>Always set the <code>secure</code> flag to `true` on a cookie before adding it
+<p>Always set the `secure` flag to `true` on a cookie before adding it
```
Why did you change this?
As far as I know, this will not work, but `<code>` will work.

dellalibera replied:
Sorry, I was thinking the opposite. Thanks for the feedback. I made the change.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+    abstract DataFlow::Node getCookieOptionsArgument();
+
+    /**
+     * Predicate that determines if a cookie is insecure.
+     */
+    abstract predicate isInsecure();
```
```suggestion
     * Holds if this cookie is secure.
     */
    abstract predicate isSecure();
```
Optional: I think it is clearer to flip the polarity and avoid the negation in the predicate name.
Please note that this requires a change to all implementations of this predicate below.

dellalibera replied:
Yes, I agree with you. I changed the polarity (I did it in one commit) of all the implementations of this predicate.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+
+  /**
+   * A cookie set using the `express` module `cookie-session` (https://github.com/expressjs/cookie-session).
+   * The flag `secure` is set to `false` by default for HTTP, `true` by default for HTTPS (https://github.com/expressjs/cookie-session#cookie-options).
```
```suggestion
```
The subtleties of when the cookie is secure or not are better described in the `isSecure` predicate.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+   */
+  class InsecureCookieSession extends ExpressLibraries::CookieSession::MiddlewareInstance,
+    InsecureCookies {
+    InsecureCookieSession() { this instanceof ExpressLibraries::CookieSession::MiddlewareInstance }
```
This charpred is redundant since the class already extends the `MiddlewareInstance` in the `extends` above, so we can omit it.

dellalibera replied:
Yes, you are right. I made the change

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+    override string getKind() { result = "cookie-session" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.getOption("cookie").(DataFlow::SourceNode)
```
```suggestion
      result = this.getOption("cookie")
```
This cast is redundant, the `result`-type in the signature of this predicate performs the same cast.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getOption("cookie").(DataFlow::SourceNode)
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
```
```suggestion
    private DataFlow::Node getCookieFlagValue(string flag) {
```
I think this can be private

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if the `secure` flag is explicitly set to `false`.
+    override predicate isInsecure() { getCookieFlagValue(flag()).mayHaveBooleanValue(false) }
```
```suggestion
    override predicate isInsecure() { 
        // A cookie is insecure if the `secure` flag is explicitly set to `false`.
        getCookieFlagValue(flag()).mayHaveBooleanValue(false)
    }
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+
+  /**
+   * A cookie set using the `express` module `express-session` (https://github.com/expressjs/session).
+   * The flag `secure` is not set by default (https://github.com/expressjs/session#cookiesecure).
+   * The default value for cookie options is { path: '/', httpOnly: true, secure: false, maxAge: null }.
```
As above: this description can be moved to the relevant predicate.
```suggestion
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+    override string getKind() { result = "express-session" }
+
+    override DataFlow::SourceNode getCookieOptionsArgument() {
+      result = this.getOption("cookie").(DataFlow::SourceNode)
```
```suggestion
      result = this.getOption("cookie")
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getOption("cookie").(DataFlow::SourceNode)
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
```
```suggestion
    private DataFlow::Node getCookieFlagValue(string flag) {
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
```
```suggestion
    override predicate isInsecure() {
        // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      ) and
```
```suggestion
      not getCookieFlagValue(flag()).mayHaveBooleanValue(true) and
```
`getCookieFlagValue(flag)` implies the existence of `getCookieOptionsArgument()`, so this exists can be simplified.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+   */
+  class InsecureExpressCookieResponse extends InsecureCookies {
+    InsecureExpressCookieResponse() {
+      this = any(Express::ResponseExpr response).flow().getALocalSource().getAMemberCall("cookie")
```
```suggestion
      this = any(Express::ResponseExpr response).flow().getALocalSource().getAMethodCall("cookie")
```
I do not think the following will work, so we should use `getAMethodCall`:

```js
let cookie = res.cookie;
cookie({...})
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.(DataFlow::InvokeNode).getLastArgument().getALocalSource()
+    }
+
+    DataFlow::Node getCookieFlagValue(string flag) {
```
```suggestion
    private DataFlow::Node getCookieFlagValue(string flag) {
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
```
```suggestion
    override predicate isInsecure() {
        // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.

```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      )
```
```suggestion
      not getCookieFlagValue(flag()).mayHaveBooleanValue(true)    
```
As above, I belive the implication makes the `exists` redundant.

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result.asExpr() = this.asExpr().(ArrayExpr).getAnElement()
+    }
+
+    // A cookie is insecure if the 'secure' flag is not specified in the cookie definition.
+    override predicate isInsecure() {
```
```suggestion
    override predicate isInsecure() {
        // A cookie is insecure if the 'secure' flag is not specified in the cookie definition.
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+    override predicate isInsecure() {
+      not exists(string s |
+        getCookieOptionsArgument().mayHaveStringValue(s) and
+        s.matches("%; secure%")
```
Hmm. What if the string is `...;secure...`? That is: it lacks a space between `;` and `secure`.

How about something like`s.regexpMatch("(.*;)?\\s*secure.*")` instead? That should allow any whitespace variation, and also permit the string to start with `secure`.

dellalibera replied:
Again, you are right. I missed that. I made the change

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+      result = this.getCookieOptionsArgument().getAPropertyWrite(flag).getRhs()
+    }
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
```
```suggestion
    override predicate isInsecure() {
        // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
```

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+
+    // A cookie is insecure if there are not cookie options with the `secure` flag set to `true`.
+    override predicate isInsecure() {
+      not exists(DataFlow::SourceNode cookieOptions |
+        cookieOptions = this.getCookieOptionsArgument() and
+        getCookieFlagValue(flag()).mayHaveBooleanValue(true)
+      )
```
```suggestion
      not getCookieFlagValue(flag()).mayHaveBooleanValue(true)
```
As above.

dellalibera commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.qll:
```diff
@@ -0,0 +1,161 @@
+  /**
+   * Abstract class to represent different cases of insecure cookie settings.
+   */
+  abstract class InsecureCookies extends DataFlow::Node {
```
Yes, no problem. I made the change.

dellalibera commented:
Hi @esbena,

sorry for the late reply.
I super appreciated your feedback (I learnt a lot).

About the `js-cookie` library I was inspired by the code the you mentioned [https://github.com/github/codeql/blob/71933a4d8a32d32f085f14b7b44f41d1cdcae2e4/javascript/ql/src/semmle/javascript/frameworks/CookieLibraries.qll](https://github.com/github/codeql/blob/71933a4d8a32d32f085f14b7b44f41d1cdcae2e4/javascript/ql/src/semmle/javascript/frameworks/CookieLibraries.qll).
However the existing implementation provides access only to the value of a cookie and not its options (please correct me if I am wrong).

I made all the changes that you suggested and then I also changed the polarity of the predicate `isInsecure` (I changed also all the implementations of this predicate).

Finally, I removed all the examples and I added them as tests, as you suggested (I hope I added in the correct folder).

Let me know if now it’s ok and if there is something else I can do. 
Again, thanks for all the feedback.

esbena commented:
Thanks. The new tests seem fine.
I have a few QL style comments left. 
Please remember to run `codeql query format` before pushing, otherwise the PR checks will fail.

dellalibera commented:
Thanks for your reply.  I formatted the code.
Let me know if there is something else I can do.

dellalibera commented:
@esbena, no problem :)
thanks for the other suggestions (I learnt some new stuff). 

Let me know if there is something else I can do.

intrigus-lgtm commented:
> ^ Pro tip: in the "Files changed" view, you have the option of adding review suggestions to a batch and commit them all at once. (https://github.com/github/codeql-go/pull/66#issuecomment-601748477)

esbena commented on javascript/ql/src/experimental/Security/CWE-614/InsecureCookie.ql:
```diff
@@ -0,0 +1,19 @@
+
+from Cookie cookie
+where not cookie.isSecure()
+select "Cookie is added to response without the 'secure' flag being set to true (using " +
```
```suggestion
select cookie, "Cookie is added to response without the 'secure' flag being set to true"
```

- This selects the right information for an alert query (mandatory)
- I dropped the `getKind()`, it is nice for testing, but seems redundant when looking at the concrete alerts (optional)
- You need to regenerate the expected outputs with this change (mandatory)

esbena reviewed (changes requested):
One more thing. I stumbled upon it when I looked at the results of the query on LGTM.com.

dellalibera commented:
@intrigus-lgtm thanks for the tip.

@esbena I made the change and updated the expected output.

esbena commented:
I took the liberty of pushing a formatting fix to your branch. The tests should go green shortly.

esbena commented:
Thank you for yet another contribution @dellalibera.

dellalibera commented:
@esbena It was really a pleasure. Thank you for all the feedback.
